### ISSUE TITLE
vim9 broken indentation

### ISSUE THREAD

**Issue by monkoose** [vim/vim#11023]
### Steps to reproduce

A lot of expected indentations for a programming language are currently broken

as example
- any of line continuations on the next line, especially dictionaries
- line continuations for chained method syntax `->`
- next lines after closing brackets
- and some more
`=` is not even usable and shifts code to the right.

Maybe completely new runtime files can be created for vim9 filetypes, instead of mixing them with old one? (syntax, indent).

### Expected behaviour

Correct indentation

### Version of Vim

9.0.0231

### Environment

Linux

**Comment by lacygoill** [comment]
A long time ago, I started re-writing the legacy indent plugin for Vim9 (it's [here](https://github.com/lacygoill/config/blob/master/.vim/indent/vim.vim)).  But I stopped the refactoring.  At some point, I'll try to finish it, but not any time soon; I'm working on something else.

**Comment by monkoose** [comment]
Can you elaborate what is currently missing, maybe I could help.
I'm using your vim9-syntax and it is  good enough. Thank you for your work on it.

**Comment by lacygoill** [comment]
> good highlights, but too much logic and feels slow sometimes

You can measure the most costly rules with [`:help :syntime`](https://vimhelp.org/syntax.txt.html#%3Asyntime).

I'm not sure I could improve them much.  Usually, what increases the time the most is lookbehinds.
For example, on my machine, the 2 most costly rules are `vim9Oper` and `vim9FuncCallUser`:

    1.099728   34182  5108    0.000494    0.000032  vim9Oper           \%(\_s\)\@1<=\%([-+*/%]\|\.\.\|||\|&&\|??\|?\|<<\
    1.060079   35840  5663    0.000338    0.000030  vim9FuncCallUser   \<\%([gs]:\u\w*\|\u\w*\|\%(\w\|#\)*\|:\@1<!\w\+\.

They both contain a lookbehind:

    \%(\_s\)\@1<=\%([-+*/%]\|\.\.\|||\|&&\|??\|?\|<<\
            ^---^

    \<\%([gs]:\u\w*\|\u\w*\|\%(\w\|#\)*\|:\@1<!\w\+\.
                                          ^---^

But they are limited to 1 byte (which should help).

> Can you elaborate what is currently missing, maybe I could help.

I guess short tests would help.  Without it's hard to tell if by fixing something you didn't break something else.  If there are small bugs to fix, which don't require too much time, I guess I could fix them.  But if it needs more work, I won't be able to do it for a few months (I need to finish [this book](https://www.packtpub.com/product/learn-python-programming/9781801815093) and [this one](https://www.packtpub.com/product/python-object-oriented-programming/9781801077262) so I can test [`:help vim9-classes`](https://vimhelp.org/vim9.txt.html#vim9-classes)).

If you could start by posting a few short snippets of code, with expected indentation and actual indentation, I could try to work on them when I have time.

**Comment by monkoose** [comment]
> a few short snippets of code

Can you provide some example how they should look. Like some correct indentation and current (based on what default vim indent file or this one of yours)?

**Comment by lacygoill** [comment]
> Can you provide some example how they should look. Like some correct indentation of some code structure paired with current incorrect one(based on what default vim indent file or this one of yours)? Or something else?

I guess we could start with some short code blocks like these:

    augroup TestSth
       for i in []
       endfor
    augroup END

    def Func()
        var d = dd
        ->extend({
        })
        eval 0
    enddef

    if 1
        global/^/echo '' | if true | eval 1 + 2 | endif
        global/^/echo '' | if true | eval 1 + 2 | endif

        command Test echo '' | if true | eval 1 + 2 | endif
        command Test echo '' | if true | eval 1 + 2 | endif
    endif

If you could write tests I could include in my local copy of Vim, that would help.  You can learn how to write them [here](https://github.com/vim/vim/blob/master/runtime/indent/testdir/README.txt).  Basically, you would need to write a `.in` file containing the code to be indented (set the indentation level of all the lines to 0), and a `.ok` file containing the code as it should be indented.

**Comment by monkoose** [comment]
> You can learn how to write them [here](https://github.com/vim/vim/blob/master/runtime/indent/testdir/README.txt).

Thanks that is the way. It would take some time though. Will update this issue when i will come up with something.

> You will have to give reproducible examples of what you call "broken".

Bram, I will follow @lacygoill advice and will provide some indent test samples for what is not working.
But it is literally like a lot of stuff currently not indented at all(from user perspective it is like nothing was done for vim9 in indent file).

**Comment by brammool** [comment]
> I guess we could start with some short code blocks like these:
> 
>     augroup TestSth
>        for i in []
>        endfor
>     augroup END

This looks OK.

>     def Func()
>         var d = dd
>         ->extend({
>         })
>         eval 0
>     enddef

Here "eval 0" is not indented sufficient.
The line starting with "->" would better be indented more, like a
continuation line.

>     if 1
>         global/^/echo '' | if true | eval 1 + 2 | endif
>         global/^/echo '' | if true | eval 1 + 2 | endif
> 
>         command Test echo '' | if true | eval 1 + 2 | endif
>         command Test echo '' | if true | eval 1 + 2 | endif
>     endif

Looks OK.


-- 
OLD WOMAN: Well, how did you become king, then?
ARTHUR: The Lady of the Lake, her arm clad in the purest shimmering samite,
        held Excalibur aloft from the bosom of the water to signify by Divine
        Providence ...  that I, Arthur, was to carry Excalibur ...  That is
        why I am your king!
                 "Monty Python and the Holy Grail" PYTHON (MONTY) PICTURES LTD

 /// Bram Moolenaar -- ***@***.*** -- http://www.Moolenaar.net   \\\
///                                                                      \\\
\\\        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by monkoose** [comment]
@lacygoill you can't have dictionary closing curly bracket on a new line by specification, no?

**Comment by lacygoill** [comment]
AFAIK, we can close a dictionary however we want.

There exists a limitation which is specified at [`:help command-block`](https://vimhelp.org/vim9.txt.html#command-block), but it's only for a dictionary inside a block.

**Comment by monkoose** [comment]
Some of below samples literally taken from `:help vim9` where in examples they have correct indentation. It's not a full list just some annoyances that bothers at least myself.

I am not sure how continuation lines should be indented. What i have did in `ok` file is indented them by 2 tabs, except of list and dictionaries which ends with `,` them i aligned with item of the previous line. Correct me if this is wrong.

 <details>
<summary>vim.in</summary>

```vim
# vim: ft=vim sw=4 :

vim9script

# START_INDENT

var result = Func(
arg1,
arg2
)

var result = Func(arg1,
arg2)

filter(list, (k, v) =>
v > 0)

filter(list, (k, v) => {
const x = get(list, k, 0)
return x > 0
}

if x > 0
filter(list, (k, v) => {
const x = get(list, k, 0)
return x > 0
}
endif

{
var temp = 'temp'
}

var text = lead
.. middle
.. end

var text = lead ..
middle ..
end

var total = start +
end -
correction

var result = start
:+ print

var result = positive
? PosFunc(arg)
: NegFunc(arg)

var result = GetBuilder()
->BuilderSetWidth(333)
->BuilderSetHeight(777)
->BuilderBuild()

var result = MyDict
.member

autocmd BufNewFile *.match if condition
|   echo 'match'
| endif

set cpo+=C
var lines =<< trim END
| this works
END
set cpo-=C

syn region Text
\ start='foo'
#\ comment
\ end='bar'

au CursorHold * echom 'BEFORE bar'
#\ some comment
| echom 'AFTER bar'

def MyFunc(text: string,
separator = '-'
): string

def MyFunc(
text: string,
separator = '-'
): string

[var1, var2] =
Func()

const list = ['one',
'two']

const list = [
'one',
'two',
]

const dict = {one: 1,
two: 2
}

const dict = {
one: 1,
two: 2
}

if true
const dict =
{
one: 1,
two: 2
}
endif

def Func()
return {
one: 1
}
enddef

# END_INDENT
```
</details>

<details>
<summary>vim.ok</summary>

```vim
# vim: ft=vim sw=4 :

vim9script

# START_INDENT

var result = Func(
    arg1,
    arg2
)

var result = Func(arg1,
        arg2)

filter(list, (k, v) =>
        v > 0)

filter(list, (k, v) => {
    const x = get(list, k, 0)
    return x > 0
}

if x > 0
filter(list, (k, v) => {
    const x = get(list, k, 0)
    return x > 0
}
endif

{
    var temp = 'temp'
}

var text = lead
        .. middle
        .. end

var text = lead ..
        middle ..
        end

var total = start +
        end -
        correction

var result = start
:+ print

var result = positive
        ? PosFunc(arg)
        : NegFunc(arg)

var result = GetBuilder()
                ->BuilderSetWidth(333)
                ->BuilderSetHeight(777)
                ->BuilderBuild()

var result = MyDict
                .member

autocmd BufNewFile *.match if condition
        |   echo 'match'
        | endif

set cpo+=C
var lines =<< trim END
        | this works
END
set cpo-=C

syn region Text
        \ start='foo'
        #\ comment
        \ end='bar'

au CursorHold * echom 'BEFORE bar'
        #\ some comment
        | echom 'AFTER bar'

def MyFunc(text: string,
           separator = '-'): string

def MyFunc(
    text: string,
    separator = '-'
): string

[var1, var2] =
        Func()

const list = ['one',
              'two']

const list = [
    'one',
    'two',
]

const dict = {one: 1,
              two: 2}

const dict = {
    one: 1,
    two: 2
}

if true
    const dict =
            {
                one: 1,
                two: 2
            }
endif

def Func()
    return {
        one: 1
    }
enddef

# END_INDENT
```
</details>

**Comment by brammool** [comment]
> AFAIK, we can close a dictionary however we want.
> 
> There exists a limitation which is specified at [`:help
> command-block`](https://vimhelp.org/vim9.txt.html#command-block), but
> it's only for a dictionary inside a block.

We cannot handle everything, but we should be able to handle the common
cases:  "}" at the start of a line and "}" at the end of a line.  It
gets more tricky if there is nesting, e.g. in a function call argument:

	TheFunc({
		key: value })

Using searchpair() it can find the matching "(", then the "{}" is
irrelevant.


-- 
BLACK KNIGHT: None shall pass.
ARTHUR:       I have no quarrel with you, brave Sir knight, but I must cross
              this bridge.
BLACK KNIGHT: Then you shall die.
                 "Monty Python and the Holy Grail" PYTHON (MONTY) PICTURES LTD

 /// Bram Moolenaar -- ***@***.*** -- http://www.Moolenaar.net   \\\
///                                                                      \\\
\\\        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by brammool** [comment]
> If we start with `vim --clean` and press `gg=G` in this file, the
> indentation gets wrong after the `Func()` header.
> 
> FWIW, I think we need more tests, and an indent plugin completely
> dedicated to Vim9.  I'll work on it, but that'll take time.

That will be good.  I had a look at the first examples and realise we
need a better line continuation detection.  Something like a line ending
in one of "[{[" or the next line starting with "->".  That is not too
difficult, but then we also need a mechanism to align below a number of
continued lines, thus it has to go back to the first one.  That's more
complicated.  Add to that avoiding to use matches in a comment.

	var v = (       -> trailing "(" starts line continuation
	  3 + 4		-> nothing special
	  )		-> end of expression indicates continued line
	var x: number	-> needs to align with previous "var"

And then:
	for v in [
	  3, 4
	  ]
	  var x: number  -> align with "for" plus a 'shiftwidth'

This is much more complicated than legacy script.  Still, we can
probably use the same rules in legacy script, these constructs just
won't appear there.

-- 
DENNIS: You can't expect to wield supreme executive power just 'cause some
        watery tart threw a sword at you!
                 "Monty Python and the Holy Grail" PYTHON (MONTY) PICTURES LTD

 /// Bram Moolenaar -- ***@***.*** -- http://www.Moolenaar.net   \\\
///                                                                      \\\
\\\        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by lacygoill** [comment]
The `vim.in` file given earlier is missing 2 closing parens (and maybe 2 `enddef`s):
```diff
--- /tmp/t.vim	2022-09-07 22:46:08.770850389 +0200
+++ /tmp/a.vim	2022-09-07 22:46:32.802750534 +0200
@@ -18,13 +18,13 @@
 filter(list, (k, v) => {
 const x = get(list, k, 0)
 return x > 0
-}
+})
 
 if x > 0
 filter(list, (k, v) => {
 const x = get(list, k, 0)
 return x > 0
-}
+})
 endif
 
 {
@@ -80,11 +80,13 @@
 def MyFunc(text: string,
 separator = '-'
 ): string
+enddef
 
 def MyFunc(
 text: string,
 separator = '-'
 ): string
+enddef
 
 [var1, var2] =
 Func()
```
---

The PR https://github.com/vim/vim/pull/11079 should pass all the existing tests (provided that we change 2 lines in the `vim.ok` file; the changes are sensible).  I think it also passes all the examples posted in this thread.

More work needs to be done though.  In particular, comments inside multiline constructs.  Expected:

    echo 1
        # some comment
        + 2
        # some comment
        + 3

Actual:

    echo 1
    # some comment
        + 2
    # some comment
        + 3

**Comment by lacygoill** [comment]
> Currently all the examples are broken except of :+ print that was added just for showing how it is different from + print.

The new plugin should handle all of them correctly.

This is the diff for `vim.in`:
<details>
<summary>diff</summary>

```diff
--- /tmp/before	2022-09-08 02:02:20.238487541 +0200
+++ /tmp/after	2022-09-08 02:02:04.542539721 +0200
@@ -3,15 +3,15 @@
 # START_INDENT
 
 var result = Func(
-arg1,
-arg2
+    arg1,
+    arg2
 )
 
 var result = Func(arg1,
-arg2)
+    arg2)
 
 filter(list, (k, v) =>
-v > 0)
+    v > 0)
 
 filter(list, (k, v) => {
     const x = get(list, k, 0)
@@ -30,23 +30,23 @@
 }
 
 var text = lead
-.. middle
-.. end
+    .. middle
+    .. end
 
 var text = lead ..
-middle ..
-end
+    middle ..
+    end
 
 var total = start +
-end -
-correction
+    end -
+    correction
 
 var result = start
 :+ print
 
 var result = positive
-? PosFunc(arg)
-: NegFunc(arg)
+    ? PosFunc(arg)
+    : NegFunc(arg)
 
 var result = GetBuilder()
     ->BuilderSetWidth(333)
@@ -54,51 +54,51 @@
     ->BuilderBuild()
 
 var result = MyDict
-.member
+    .member
 
 autocmd BufNewFile *.match if condition
-|   echo 'match'
-| endif
+    |   echo 'match'
+    | endif
 
 set cpo+=C
 var lines =<< trim END
-| this works
+    | this works
 END
 set cpo-=C
 
 syn region Text
     \ start='foo'
-#\ comment
+    #\ comment
     \ end='bar'
 
 au CursorHold * echom 'BEFORE bar'
-#\ some comment
-| echom 'AFTER bar'
+    #\ some comment
+    | echom 'AFTER bar'
 
 def MyFunc(text: string,
-    separator = '-'
-    ): string
+        separator = '-'
+        ): string
 enddef
 
 def MyFunc(
-    text: string,
-    separator = '-'
-    ): string
+        text: string,
+        separator = '-'
+        ): string
 enddef
 
 [var1, var2] =
-Func()
+    Func()
 
 const list = ['one',
-'two']
+    'two']
 
 const list = [
     'one',
     'two',
-    ]
+]
 
 const dict = {one: 1,
-two: 2
+    two: 2
 }
 
 const dict = {
@@ -108,10 +108,10 @@
 
 if true
     const dict =
-    {
+        {
             one: 1,
             two: 2
-    }
+        }
 endif
 
 def Func()
```
</details>

---

   > We cannot handle everything, but we should be able to handle the common
   > cases:  "}" at the start of a line and "}" at the end of a line.  It
   > gets more tricky if there is nesting, e.g. in a function call argument:
   > 
   > 	TheFunc({
   > 		key: value })
   > 
   > Using searchpair() it can find the matching "(", then the "{}" is
   > irrelevant.

The new indent plugin correctly handles that.

Current indent plugin:
```vim
vim9script
TheFunc({
key: value })
```
New one:
```vim
vim9script
TheFunc({
    key: value })
```
---

   > That will be good.  I had a look at the first examples and realise we
   > need a better line continuation detection.  Something like a line ending
   > in one of "[{[" or the next line starting with "->".  That is not too
   > difficult, but then we also need a mechanism to align below a number of
   > continued lines, thus it has to go back to the first one.  That's more
   > complicated.  Add to that avoiding to use matches in a comment.
   >
   >         var v = (       -> trailing "(" starts line continuation
   >           3 + 4		-> nothing special
   >           )		-> end of expression indicates continued line
   >         var x: number	-> needs to align with previous "var"

Same thing.  Current plugin:
```vim
vim9script
var v = (
3 + 4
)
var x: number
```
New one:
```vim
vim9script
var v = (
    3 + 4
)
var x: number
```
Although, it doesn't indent the closing paren.  Instead, it's aligned with the `var` statement, which seems reasonable.

---

   > And then:
   >         for v in [
   >           3, 4
   >           ]
   >           var x: number  -> align with "for" plus a 'shiftwidth'

Both the current plugin and the new one correctly handle that:
```vim
vim9script
for v in [
        3, 4
        ]
    var x: number
endfor
```
---

   > I had more code when I stopped working on the refactoring:

Same thing, the new plugin correctly handles all of it:
<details>
<summary>diff</summary>

```diff
--- /tmp/before	2022-09-08 02:26:13.170037718 +0200
+++ /tmp/after	2022-09-08 02:26:24.489975223 +0200
@@ -2,101 +2,101 @@
 
 #     {
 
-    #     var d = {
-        #     }
+#     var d = {
+#     }
+
+#     var d = {
+#         k: '
+#     }
+
+#     var d = {
+#         k: [],
+#     }
+
+augroup TestSth
+    for i in []
+    endfor
+augroup END
+
+def Func(): dict<any>
+    return {
+        a: 1,
+        b: 2,
+        c: 3,
+    }
+enddef
+
+def Func(): dict<any>
+    if true
+        return {
+            a: 1,
+            b: 2,
+            c: 3,
+        }
+    else
+        return {
+            a: 1,
+            b: 2,
+            c: 3,
+        }
+    endif
+enddef
+
+if 1
+    echo { a: 1 }
+    echo ''
+endif
 
-        #     var d = {
-            #         k: '
-            #     }
-
-            #     var d = {
-                #         k: [],
-                #     }
-
-                augroup TestSth
-                    for i in []
-                    endfor
-                augroup END
-
-                def Func(): dict<any>
-                    return {
-                        a: 1,
-                        b: 2,
-                        c: 3,
-                    }
-                enddef
-
-                def Func(): dict<any>
-                    if true
-                        return {
-                            a: 1,
-                            b: 2,
-                            c: 3,
-                        }
-                    else
-                        return {
-                            a: 1,
-                            b: 2,
-                            c: 3,
-                        }
-                    endif
-                enddef
-
-                if 1
-                    echo { a: 1 }
-                    echo ''
-                endif
-
-
-                {
-                        eval 1 + 0
-                        eval 2 + 0
-                        eval 3 + 0
-                }
-
-                def Func()
-                    if !![
-                        {key1: 'value1'},
-                        {key2: 'value2'},
-                        {key3: 'value3'},
-                        ]
-                    eval 1 + 0
-                    eval 2 + 0
-                    eval 3 + 0
-                endif
-            enddef
-            Func()
-
-            def Func()
-                for x in [
-                        {key1: 'value1'},
-                        {key2: 'value2'},
-                        {key3: 'value3'},
-                        ]
-                    eval 1 + 0
-                    eval 2 + 0
-                    eval 3 + 0
-                endfor
-            enddef
-            Func()
-
-            def Func()
-                var d = dd
-                    ->extend({
-                    })
-                eval 0
-            enddef
-
-            augroup Name
-                autocmd WinLeave * if true | eval 1 + 2 | endif
-                autocmd WinLeave * if true | eval 1 + 2 | endif
-            augroup END
-
-            if 1
-                global/^/if true | eval 1 + 2 | endif
-            global/^/if true | eval 1 + 2 | endif
 
-        command Test if true | eval 1 + 2 | endif
+{
+    eval 1 + 0
+    eval 2 + 0
+    eval 3 + 0
+}
+
+def Func()
+    if !![
+            {key1: 'value1'},
+            {key2: 'value2'},
+            {key3: 'value3'},
+            ]
+        eval 1 + 0
+        eval 2 + 0
+        eval 3 + 0
+    endif
+enddef
+Func()
+
+def Func()
+    for x in [
+            {key1: 'value1'},
+            {key2: 'value2'},
+            {key3: 'value3'},
+            ]
+        eval 1 + 0
+        eval 2 + 0
+        eval 3 + 0
+    endfor
+enddef
+Func()
+
+def Func()
+    var d = dd
+        ->extend({
+        })
+    eval 0
+enddef
+
+augroup Name
+    autocmd WinLeave * if true | eval 1 + 2 | endif
+    autocmd WinLeave * if true | eval 1 + 2 | endif
+augroup END
+
+if 1
+    global/^/if true | eval 1 + 2 | endif
+    global/^/if true | eval 1 + 2 | endif
+
+    command Test if true | eval 1 + 2 | endif
     command Test if true | eval 1 + 2 | endif
 endif
 
```
</details>

---

Note that when a line B continues from a previous one A, if A opens a block, B is indented twice:
```vim
vim9script
for i in [1] +
        [2]
    echo i
endfor
```
Here, A is the first line of the `for` statement, and B is the `[2]` line.

Indenting twice makes it easier to spot where the start of the block ends and when its body starts.  If B was indented only once, the previous snippet would be indented like this:
```vim
vim9script
for i in [1] +
    [2]
    echo i
endfor
```
Here, a careless reader might wrongly think that `[2]` is part of the loop's body.
This also applies to function's headers:
```vim
vim9script
def Func(
        n: number,
        s: string
        )
    echo n
    echo s
enddef
```
Notice that the lines declaring the `n` and `s` parameters are indented twice.
If they were indented only once, it would give:
```vim
vim9script
def Func(
    n: number,
    s: string
    )
    echo n
    echo s
enddef
```
Which is again more ambiguous because the parameters are indented just like the statements inside the body.

**Comment by lacygoill** [comment]
I guess this issue can be closed now.  The new script was merged in the [last update of the runtime files](https://github.com/vim/vim/commit/9fbdbb814f4ad67a14979aba4a6a49800c2f1a99).  If there are still bugs, new issue reports can be opened.
